plugins/inputs/docker_log/docker_logs.go

```diff
@@ -0,0 +1,440 @@
+	return nil
+}
+
+/*Following few functions have been inherited from telegraf docker input plugin*/
+func (d *DockerLogs) createContainerFilters() error {
+	filter, err := filter.NewIncludeExcludeFilter(d.ContainerInclude, d.ContainerExclude)
+	if err != nil {
+		return err
+	}
+	d.containerFilter = filter
+	return nil
+}
+
+func (d *DockerLogs) createLabelFilters() error {
+	filter, err := filter.NewIncludeExcludeFilter(d.LabelInclude, d.LabelExclude)
+	if err != nil {
+		return err
+	}
+	d.labelFilter = filter
+	return nil
+}
+
+func (d *DockerLogs) createContainerStateFilters() error {
+	if len(d.ContainerStateInclude) == 0 && len(d.ContainerStateExclude) == 0 {
+		d.ContainerStateInclude = []string{"running"}
+	}
+	filter, err := filter.NewIncludeExcludeFilter(d.ContainerStateInclude, d.ContainerStateExclude)
+	if err != nil {
+		return err
+	}
+	d.stateFilter = filter
+	return nil
+}
+
+func (d *DockerLogs) addToContainerList(containerId string, logReader io.ReadCloser) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.containerList[containerId] = logReader
+	return nil
+}
+
+func (d *DockerLogs) removeFromContainerList(containerId string) error {
+	delete(d.containerList, containerId)
+	return nil
+}
+
+func (d *DockerLogs) containerInContainerList(containerId string) bool {
+	if _, ok := d.containerList[containerId]; ok {
+		return true
+	}
+	return false
+}
+
+func (d *DockerLogs) stopAllReaders() error {
+	for _, container := range d.containerList {
+		container.Close()
+	}
+	return nil
+}
+
+func (d *DockerLogs) containerListUpdate(acc telegraf.Accumulator) error {
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout.Duration)
+	defer cancel()
+	if d.client == nil {
+		log.Printf("%s : Dock client is null", ERR_PREFIX)
+		return nil
+	}
+	containers, err := d.client.ContainerList(ctx, d.opts)
+	if err != nil {
+		log.Printf("%s : %s ", ERR_PREFIX, err.Error())
+		return err
+	}
+	for _, container := range containers {
+		if d.containerInContainerList(container.ID) {
+			continue
+		}
+		d.wg.Add(1)
+		/*Start a new goroutine for every new container that has logs to collect*/
+		go func(c types.Container) {
+			defer d.wg.Done()
+			err := d.getContainerLogs(c, acc)
+			if err != nil {
+				log.Printf("%s : %s ", ERR_PREFIX, err.Error())
+			}
+			d.removeFromContainerList(c.ID)
+			return
+		}(container)
+	}
+	return nil
+}
+
+func (d *DockerLogs) getContainerLogs(
+	container types.Container,
+	acc telegraf.Accumulator,
+) error {
+	c, err := d.client.ContainerInspect(context.Background(), container.ID)
+	if err != nil {
+		log.Printf("%s : %s", ERR_PREFIX, err.Error())
+		return err
+	}
+	logOptions := types.ContainerLogsOptions{
+		ShowStdout: true,
+		ShowStderr: true,
+		Timestamps: false,
+		Details:    true,
+		Follow:     true,
+		Tail:       "0",
+	}
+	logReader, err := d.client.ContainerLogs(context.Background(), container.ID, logOptions)
+	d.addToContainerList(container.ID, logReader)
```

Comment:
Suggest moving this into the `containerListUpdate` where the rest of the related functions are used.  Then we will have add/remove/in functions all in one place.

Must lock d.containerList for this action, since it is called within a goroutine that is not waited for within Gather.

Reply:
1. Moved addToContainerList to containerListUpdate and passed logReader into tailContainerLogs.
2. addToContainerList already has a mutex lock for d.containerList  . Is there anything else I should be adding?